### DISCUSSION TITLE
Open Generic as a type parameter

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#6290] [Q&A]
Let's say I have the following code:
```
public abstract class ChildBase {
  public DateTimeOffset Date {get; set;}
}

public class Parent<TChild> where TChild : ChildBase, new() {
  public TChild Child {get; set;} = new();
}

```

Given the above, ```Parent.Child.Date``` is always valid to get or set.

How would I write the following code?
```
object Input = SomeMethod();
if(Input is Parent<> {} V1) {
  V1.Child.Date = DateTimeOffset.UtcNow;
}
```

Thanks!

**Comment by svick** [comment]
Effectively, what you're asking for is covariance. As you probably know, you can use covariance if you change `Parent` to:

```c#
public interface IParent<out TChild> where TChild : ChildBase {
  TChild Child { get; }
}
```

Then your condition would be:

```c#
object Input = SomeMethod();
if (Input is IParent<ChildBase> V1) {
  V1.Child.Date = DateTimeOffset.UtcNow;
}
```

[Full code.](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6D2AjAVgUwGNgACASSgAcJSBeEgZUwFt8BZfYAC0wBMAKAJQBudAEsAZiX4VqpMQGdyABQCGYfFGAAeAMJcxAG14AhVQvwA+EgDUkgkgG90JW0gB0+o73cARVcD4JPT+gQAqYqwA8hISFsDuAKrAhABymADuomgAvuhYeESkTKwc3HxCwdZQ+Bkkahpa2iX4XsaWQtnocADMJKrYCsBgqsQkiCRtpuZBzmiuvSSh+BHRsfFLAbMkLq57AOacVSS8+BKqEIbA2Xt7G7TWcEgAnPwAJABEDJzAYlD7m0CJGAmCcADdVIYIPgch8RLs8rl8mhFn9AmBzoQgmQGpodJgaCQwlNrBkuPgNESpiQQJMDMYzBYnLtifTeHTvE4SIdriREYjugh6uo8dpWd5rLScSKmuL2iQyRSgnL2bSpoz8AAaEg1DKVOa3RYqjnGLk84QkeIWnLBHW1TroAUooUtalqtkark5IA)

If you can't change `Parent` like this, then I think your best option is reflection.

**Comment by Richiban** [comment]
Actually I think what's closer to what the OP wants is existential types:

Given

```cs
public interface IParent
{
	type TChild : ChildBase;

	TChild Child { get; }
}
```

You could then perform a regular type check (because the generic type doesn't form part of the parent type directly) and the following simply becomes legal:

```cs
if (SomeMethod() is IParent v1)
{
	v1.Child.Date = DateTimeOffset.UtcNow;
}
```

**Reply by 333fred** [reply]
I don't think this is likely to be a solution, as I don't think IParent will be directly usable there if it has an existential type.
